File: projeto/estudo-dirigido-1.2/lib/questao02/produto.dart
class Produto {
  String _nome;
  double _preco;
  int _estoque;

  Produto(this._nome, this._preco, this._estoque) {
    _validar();
  }

  // Método privado de validação
  void _validar() {
    if (_nome.trim().isEmpty) {
      print("O nome não pode ser vazio.");
    }
    if (_preco < 0) {
      print("O preço não pode ser negativo.");
    }
    if (_estoque < 0) {
      print("O estoque não pode ser negativo.");
    }
  }

  // Getters
  String get nome => _nome;
  double get preco => _preco;
  int get estoque => _estoque;

  // Setters
  set nome(String valor) {
    String novoNome = valor;
    if (novoNome.trim().isEmpty) {
      print("O nome não pode ser vazio.");
    }
    _nome = novoNome;
  }

  set preco(double valor) {
    if (valor < 0) {
      print("O preço não pode ser negativo.");
    }
    _preco = valor;
  }

  set estoque(int valor) {
    if (valor < 0) {
      print("O estoque não pode ser negativo.");
    }
    _estoque = valor;
  }

  void repor(int qtd) {
    if (qtd < 0) {
      print("Quantidade para reposição deve ser positiva.");
      return;
    }
    _estoque += qtd;
  }

  void vender(int qtd) {
    if (qtd <= 0) {
      print("Quantidade para venda deve ser positiva.");
      return;
    }
    if (qtd <= _estoque) {
      _estoque -= qtd;
    } else {
      print("Estoque insuficiente");
    }
  }

  void exibirFicha() {
    print("Nome: $_nome | Preço: $_preco | Estoque: $_estoque");
  }
}

File: projeto/estudo-dirigido-1.2/lib/questao02/__produto__.py
class Produto:
    def __init__(self, nome: str, preco: float, estoque: int):
        self._nome = nome
        self._preco = preco
        self._estoque = estoque
        self._validar()  # valida após a inicialização

    def _validar(self):
        """Método privado de validação de todos os campos."""
        if not self._nome or self._nome.strip() == "":
            raise ValueError("O nome não pode ser vazio.")
        if self._preco < 0:
            raise ValueError("O preço não pode ser negativo.")
        if self._estoque < 0:
            raise ValueError("O estoque não pode ser negativo.")

    # Getters e Setters no estilo Python (property)
    @property
    def nome(self) -> str:
        return self._nome

    @nome.setter
    def nome(self, valor: str):
        if not valor or valor.strip() == "":
            raise ValueError("O nome não pode ser vazio.")
        self._nome = valor

    @property
    def preco(self) -> float:
        return self._preco

    @preco.setter
    def preco(self, valor: float):
        if valor < 0:
            raise ValueError("O preço não pode ser negativo.")
        self._preco = valor

    @property
    def estoque(self) -> int:
        return self._estoque

    @estoque.setter
    def estoque(self, valor: int):
        if valor < 0:
            raise ValueError("O estoque não pode ser negativo.")
        self._estoque = valor

    # Métodos de negócio
    def repor(self, qtd: int):
        if qtd < 0:
            print("Quantidade para reposição deve ser positiva.")
            return
        self._estoque += qtd

    def vender(self, qtd: int):
        if qtd <= 0:
            print("Quantidade para venda deve ser positiva.")
            return
        if qtd <= self._estoque:
            self._estoque -= qtd
        else:
            print("Estoque insuficiente")

    def exibir_ficha(self):
        print(f"Nome: {self._nome} | Preço: {self._preco} | Estoque: {self._estoque}")


